File: languages/js/src/entity/loader.ts
import * as proto from "../bridge/proto/availability_entities";

export interface Loader {
  type: proto.DataAvailabilityType;
  args: proto.LoaderArgs;
  toProto(): proto.Loader;
}

export class LoaderArgs {
  id: string;
  constructor(id: string) {
    this.id = id;
  }

  public toProto(): proto.LoaderArgs {
    return proto.LoaderArgs.fromPartial({ id: this.id });
  }
}

export class HostedLoader implements Loader {
  type: proto.DataAvailabilityType;
  args: LoaderArgs;

  constructor(id: string) {
    this.type = proto.DataAvailabilityType.HOSTED;
    this.args = new LoaderArgs(id);
  }

  public toProto(): proto.Loader {
    return proto.Loader.fromPartial({
      type: this.type,
      args: this.args.toProto()
    });
  }
}

export class IpfsLoader implements Loader {
  type: proto.DataAvailabilityType;
  args: LoaderArgs;

  constructor(id: string) {
    this.type = proto.DataAvailabilityType.IPFS;
    this.args = new LoaderArgs(id);
  }

  public toProto(): proto.Loader {
    return proto.Loader.fromPartial({
      type: this.type,
      args: this.args.toProto()
    });
  }
}

File: languages/go/entity/loader.go
package entity

import "github.com/bloock/bloock-sdk-go/v2/internal/bridge/proto"

type Loader interface {
	ToProto() *proto.Loader
}

type LoaderArgs struct {
	Id string
}

func (e LoaderArgs) ToProto() *proto.LoaderArgs {
	return &proto.LoaderArgs{
		Id: e.Id,
	}
}

type HostedLoader struct {
	Type proto.DataAvailabilityType
	Args LoaderArgs
}

func NewHostedLoader(id string) HostedLoader {
	return HostedLoader{
		Type: proto.DataAvailabilityType_HOSTED,
		Args: LoaderArgs{
			Id: id,
		},
	}
}

func (e HostedLoader) ToProto() *proto.Loader {
	return &proto.Loader{
		Type: e.Type,
		Args: e.Args.ToProto(),
	}
}

type IpfsLoader struct {
	Type proto.DataAvailabilityType
	Args LoaderArgs
}

func NewIpfsLoader(hash string) IpfsLoader {
	return IpfsLoader{
		Type: proto.DataAvailabilityType_IPFS,
		Args: LoaderArgs{
			hash,
		},
	}
}

func (e IpfsLoader) ToProto() *proto.Loader {
	return &proto.Loader{
		Type: e.Type,
		Args: e.Args.ToProto(),
	}
}
